### DISCUSSION TITLE
Nuxt3 - Server-side sourcemaps?

### DISCUSSION THREAD

**Discussion by coryvirok** [nuxt/nuxt#16529] [Questions] [answered]
I've been trying to get inline sourcemaps generated for my Nuxt 3 dev server in order to debug my Nuxt SSR code and get proper filenames in server-side traces. 

**TL;DR**: How can I get server-side code to include sourcemaps, (ideally inlined?)

_nuxt.config.js_

```js
import { defineNuxtConfig } from 'nuxt3'

const config = defineNuxtConfig({
  typescript: {
    strict: true,
  },
  css: ['@/assets/scss/main.scss'],
  build: {
    postcss: {
      postcssOptions: {
        plugins: {
          tailwindcss: {},
          autoprefixer: {},
        },
      },
    },
  },
  modules: ['@pinia/nuxt'],

  // no effect
  dev: true,       

  // I've also tried `vite: false` but that causes https://github.com/nuxt/framework/issues/2786
  vite: {   
    build: {
      sourcemap: true,   // no effect
    },
  },
  vue: {
    compilerOptions: {
      sourceMap: true,    // no effect
    },
  },
})
export default config

```

Relevant deps

```
@nuxt/nitro@3.0.0-27376999.6abf75d
@nuxt/vite-builder@3.0.0-27376999.6abf75d
@vitejs/plugin-vue@2.0.1
vite@2.7.13
rollup@2.65.0
nuxt3@3.0.0-27376999.6abf75d
```

I see that client-side (inline) source maps are being generated by the dev server and the code in _.nuxt/dist/nitro_ has a source map (file) as well. However, the code in _.nuxt/dist/server_ does not seem to have any sourcemaps. Rather, _server.mjs_ has `sourceMappingURL=` lines pointing to files that are nowhere to be found.

E.g. 

```
.nuxt/
  dist/
    server/
      client.manifest.json   <-- not relevant
      client.manifest.mjs    <-- no sourceMappingURL lines but this code isn't relevant
      server.mjs             <-- has sourceMappingURL lines pointing to files that I cannot find, eg. 
                             <-- //# sourceMappingURL=portal.esm.js.map
    nitro/
      index.mjs              <!-- has //# sourceMappingURL=index.mjs.map
      index.mjs.map          <-- sourcemap for index.mjs

```

**Comment by danielroe** [comment, marked as answer]
This should be possible when https://github.com/nuxt/framework/pull/2795 is merged.

**Reply by coryvirok** [reply]
Excellent! Thanks

**Reply by ayalon** [reply]
The merge request above is merged. But still I cannot see any source maps on the server.mjs. Any hints? Did it work for  you @coryvirok ?

**Reply by danielroe** [reply]
You set it like this:

```js
export default defineNuxtConfig({
  sourcemap: {
    server: true,
    client: true
  },
  // and if you need to customise the _kind_ of sourcemap for the vite build
  // you can do this in addition to the config above
  vite: {
    build: {
      sourcemap: 'inline'
    }
  }
})
```

**Comment by Hecatron** [comment]
This was something I looked into recently, there's a few undocumented hooks under Nuxt 3
For changing the vite config we can use the vite:extendConfig hook.

I think the sourcemap defaults to true for both client and server when in development mode
The below changes the server mode to inline.

```js
  // https://github.com/nuxt/framework/issues/1954
  hooks: {
    'vite:extendConfig': function (clientConfig, { isClient, isServer }) {
      if (clientConfig.mode === 'development') {
        if (isClient)
          clientConfig.build.sourcemap = true // defaults to true in development
        if (isServer)
          clientConfig.build.sourcemap = 'inline' // defaults to true in development
      }
    },
  },
```

Just for info this is an example of a launch.json file I'm using under vscode
```js
{
    "version": "0.2.0",
    "configurations": [
      {
        "name": "Launch Browser Chrome",
        "request": "launch",
        "type": "chrome",
        "url": "http://localhost:3000",
        "webRoot": "${workspaceFolder}",
      },
      {
        "name": "Launch Browser Firefox",
        "request": "launch",
        "type": "firefox",
        "url": "http://localhost:3000",
        "webRoot": "${workspaceFolder}"
      },
      {
        "name": "nuxt-server",
        "request": "launch",
        "type": "node",
        "cwd": "${workspaceFolder}",
        "runtimeExecutable": "yarn",
        "runtimeArgs": ["dev"],
        "console": "integratedTerminal"
      },
    ],
    "compounds": [
      {
        "name": "Server/Client",
        "configurations": ["nuxt-server", "Launch Browser Chrome"],
        "stopAll": true
      }
    ]
}
```

Although I did find an issue when debugging client side with the content module enabled

  * https://github.com/nuxt/content/issues/1639

**Reply by danielroe** [reply]
See https://github.com/nuxt/framework/discussions/2847#discussioncomment-3978108.

**Comment by Hecatron** [comment]
If you wanted different types of sourcemaps for client vs server then I suspect using the 'vite:extendConfig' hook might be the way to go.

That being said I'm not sure if that's something that's required if you want to debug server and client side at the same time at least with vite. I think with webpack and the old nuxt having the server side sourcemaps inline was important for some reason.
